File: calculator/src/main/java/com/turlir/calculator/member/Operators.java
package com.turlir.calculator.member;

import com.turlir.calculator.converter.Member;
import com.turlir.calculator.interpreter.NotationInterpreter;

import java.math.BigDecimal;
import java.math.MathContext;

public abstract class Operators {

    public static final Member OS = new Operator("(", 1) {

        @Override
        public void process(NotationInterpreter interpreter) {
            //
        }
    };

    public static final Operator CS = new Operator(")", 1) {

        @Override
        public void process(NotationInterpreter interpreter) {
            //
        }
    };

    public static final Member PLUS = new Operator(" + ", 2) {

        @Override
        public void process(NotationInterpreter interpreter) {
            BigDecimal a = interpreter.poolDigit();
            BigDecimal b = interpreter.poolDigit();
            interpreter.pushDigit(a.add(b));
        }
    };

    public static final Member MINUS = new Operator(" - ", 2) {

        @Override
        public void process(NotationInterpreter interpreter) {
            BigDecimal a = interpreter.poolDigit();
            BigDecimal b = interpreter.poolDigit();
            interpreter.pushDigit(b.subtract(a));
        }
    };

    public static final Member MULTI = new Operator(" * ", 3) {

        @Override
        public void process(NotationInterpreter interpreter) {
            BigDecimal a = interpreter.poolDigit();
            BigDecimal b = interpreter.poolDigit();
            interpreter.pushDigit(a.multiply(b));
        }
    };

    public static final Member DIV = new Operator(" / ", 3) {

        @Override
        public void process(NotationInterpreter interpreter) {
            BigDecimal a = interpreter.poolDigit();
            BigDecimal b = interpreter.poolDigit();
            interpreter.pushDigit(b.divide(a, MathContext.DECIMAL64));
        }
    };

    public static final Member UNARY_MINUS = new Operator("-", 4) {

        private final BigDecimal MULTIPLICAND = new BigDecimal(-1);

        @Override
        public void process(NotationInterpreter interpreter) {
            BigDecimal a = interpreter.poolDigit();
            interpreter.pushDigit(a.multiply(MULTIPLICAND));
        }
    };

    public static Member find(String token) {
        switch (token) {
            case "*":
                return MULTI;
            case "/":
                return DIV;
            case "+":
                return PLUS;
            case "-":
                return MINUS;
            case "u-":
                return UNARY_MINUS;
            case "(":
                return OS;
            case ")":
                return CS;
            default:
                throw new IllegalArgumentException("Операция не доступна");
        }
    }

}

File: calculator/src/main/java/com/turlir/calculator/member/Operators.kt
package com.turlir.calculator.member

import com.turlir.calculator.interpreter.NotationInterpreter
import java.math.BigDecimal
import java.math.MathContext

abstract class Operators {

    companion object {

        val OS = object : Operator("(", 1) {

            override fun process(interpreter: NotationInterpreter) {
                //
            }
        }

        val CS = object : Operator(")", 1) {

            override fun process(interpreter: NotationInterpreter) {
                //
            }
        }

        val PLUS = object : Operator(" + ", 2) {

            override fun process(interpreter: NotationInterpreter) {
                val a = interpreter.poolDigit()
                val b = interpreter.poolDigit()
                interpreter.pushDigit(a.add(b))
            }
        }

        val MINUS = object : Operator(" - ", 2) {

            override fun process(interpreter: NotationInterpreter) {
                val a = interpreter.poolDigit()
                val b = interpreter.poolDigit()
                interpreter.pushDigit(b.subtract(a))
            }
        }

        val MULTI = object : Operator(" * ", 3) {

            override fun process(interpreter: NotationInterpreter) {
                val a = interpreter.poolDigit()
                val b = interpreter.poolDigit()
                interpreter.pushDigit(a.multiply(b))
            }
        }

        val DIV = object : Operator(" / ", 3) {

            override fun process(interpreter: NotationInterpreter) {
                val a = interpreter.poolDigit()
                val b = interpreter.poolDigit()
                interpreter.pushDigit(b.divide(a, MathContext.DECIMAL64))
            }
        }

        val UNARY_MINUS = object : Operator("-", 4) {

            private val MULTIPLICAND = BigDecimal(-1)

            override fun process(interpreter: NotationInterpreter) {
                val a = interpreter.poolDigit()
                interpreter.pushDigit(a.multiply(MULTIPLICAND))
            }
        }

        fun find(token: String) = when (token) {
            "*" -> MULTI
            "/" -> DIV
            "+" -> PLUS
            "-" -> MINUS
            "u-" -> UNARY_MINUS
            "(" -> OS
            ")" -> CS
            else -> throw IllegalArgumentException("Операция не доступна")
        }

    }

}
